internal/controller/finalizer/types.go

```diff
@@ -0,0 +1,98 @@
+// Copyright Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the License);
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package finalizer
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/openshift-service-mesh/federation/internal/controller"
+)
+
+// FinalizeFn is a closure which can be used in the reconciler to define finalizer logic just before
+// the object is removed from kube-apiserver.
+type FinalizeFn func() error
+
+// Handler encapsulates finalizer handling. It can:
+// - add finalizer to a given object and persist it
+// - perform cleanup defined as FinalizeFn if the object is marked for deletion
+//
+// Example usage in the controller reconcile loop:
+//
+//	finalizerHandler := finalizer.NewHandler(r.Client, "federation.openshift-service-mesh.io/mesh-federation")
+//	if finalized, errFinalize := finalizerHandler.Finalize(ctx, meshFederation, func() error {
+//		// finalizer logic
+//		return nil
+//	}); finalized || errFinalize != nil {
+//		return ctrl.Result{}, errFinalize
+//	}
+//
+//	if justAdded, errAdd := finalizerHandler.Add(ctx, meshFederation); justAdded || errAdd != nil {
+//		return ctrl.Result{}, errAdd
+//	}
+type Handler struct {
+	cl            client.Client
+	finalizerName string
+}
+
+func NewHandler(cl client.Client, finalizerName string) *Handler {
+	return &Handler{
+		cl:            cl,
+		finalizerName: finalizerName,
+	}
+}
+
+// Add adds defined finalizer to the object and immediately persist it to ensure that finalizer has been added.
+// Returns true if finalizer was added and persisted in the cluster and error if the update failed.
+func (f *Handler) Add(ctx context.Context, obj client.Object) (bool, error) {
+	justAdded := controllerutil.AddFinalizer(obj, f.finalizerName)
+
+	if justAdded {
+		addFinalizer := func(saved client.Object) {
+			controllerutil.AddFinalizer(saved, f.finalizerName) // in case of conflict retry adding finalizer on the obj fetched from cluster
+		}
+		if _, errRetry := controller.RetryUpdate(ctx, f.cl, obj, addFinalizer); errRetry != nil {
+			return false, errRetry
+		}
+	}
+
+	return justAdded, nil
+}
```

Comment:
Similar to the `Finalize` method, I think using the boolean to indicate whether or not the finalizer was already present could make the usage simpler:

```
if finalizerPresent, err := finalizerHandler.Add(ctx, meshFederation); !finalizerPresent {
	return ctrl.Result{}, err
}
```

```suggestion
// Add adds the defined finalizer to the object and then persists it if the finalizer was not already present.
// Returns true if the finalizer was already present.
// Returns an error if updating the object failed.
func (f *Handler) Add(ctx context.Context, obj client.Object) (bool, error) {
	justAdded := controllerutil.AddFinalizer(obj, f.finalizerName)

	if justAdded {
		addFinalizer := func(saved client.Object) {
			controllerutil.AddFinalizer(saved, f.finalizerName) // in case of conflict retry adding finalizer on the obj fetched from cluster
		}
		_, errRetry := controller.RetryUpdate(ctx, f.cl, obj, addFinalizer)
		return false, errRetry
	}

	return true, nil
}
```

Reply:
The intention of this boolean is to distinguish if it's the first time we added the finalizer (so we updated the state in the cluster) . Therefore we need to return `justAdded` at the end to know if we should finish and wait for another reconcile or not. If we keep returning `true` we will never execute anything in reconcile which comes after piece of logic.

We need to make sure however that our predicates watch on finalizer changes (`predicate.GenerationChangedPredicate` does not, so we need one extra if we use it)

Reply:
Maybe it should be named `FinalizeIfNeeded`

Reply:
Ah ok, I missed that in the diff! That's in fact better!

I wonder `EnsureAdded` or something similar wouldn't be more helpful name. WDYT?

Reply:
> I wonder EnsureAdded or something similar wouldn't be more helpful name. WDYT?

Yeah, how about something like `AddIfNeeded` and `FinalizeIfNeeded`?